### ISSUE TITLE
Dependency calculation issues

### ISSUE THREAD

**Issue by sumerjabri** [craftercms/craftercms#6189]
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [ ] The issue is in the latest released 4.2.x
- [X] The issue is in the latest released 4.1.x
- [ ] The issue is in the latest released 4.0.x
- [ ] The issue is in the latest released 3.1.x

### Describe the issue

Dependency calculation must follow the rules:
* A hard dependency is defined as:
  1. Never-published parents (follow the parents up until /)
  2. Never-published referenced items (use the database to quickly scan by pattern)
  3. Item-specific dependencies that are never-published or have been modified (item specific means it matches the pattern configured in `studio-config.yaml`)
* A soft dependency is defined as
  1. Referenced items that are in a modified state and are not a hard dependency (to what depth?)

### Steps to reproduce

Steps:

1. Create a site from the Fitness Blueprint
2. Publish the site
3. Change the background video
4. Change the homepage
5. Try to publish the homepage and see the list of dependencies

Note the dependencies.

**Comment by jmendeza** [comment]
Some numbers tested in a large site:
342 780 items
90 605 dependencies

Retrieving dependencies for the top 10 items with more dependencies (avg 100 dependencies each).
## Items live
- API request: 120-140 ms (first request might take ~350-500 ms)
- Actual sql query: 15-20 ms (first request might take ~95-135 ms)

## Items modified
### Subsequent requests
- API request ~115-140 ms
- Actual sql query 17-23 ms

**Comment by jmendeza** [comment]
We updated the soft dependencies calculation so it gets all the dependencies regardless of item state, follows transitive dependencies and then filter by state the final list.
This allows us to get transitive dependencies following published items. e.g.:
A depends on B and B depends on C. B is published and C is modified
Before this change, publish dependencies of A would not include C.
Now it reads B and C. Then filter out B because of its state.

Because it is implemented using a [recursive CTE](https://mariadb.com/kb/en/recursive-common-table-expressions-overview/), we added a new studio config property to limit the [maximum recursion iterations](https://mariadb.com/docs/server/ref/mdb/cli/mariadbd/max-recursive-iterations/) .

This property can be set any integer between 0 and 20 (Studio-defined hard limit)
```
studio.db.maxRecursiveIterations: 10
```
cc @alhambrav
